src/drivers/net/lan91c111/lan91c111.c

```diff
@@ -252,89 +204,127 @@ static const struct net_driver lan91c111_drv_ops = {
 	.set_macaddr = lan91c111_set_macaddr
 };
 
-static irq_return_t lan91c111_int_handler(unsigned int irq_num, void *dev_id) {
+static int lan91c111_tx_pend(struct net_device *dev) {
+	struct sk_buff *skb;
+	struct smc_local *smc;
+	int ret;
+
+	smc = netdev_priv(dev);
+
+	skb = skb_queue_pop(&smc->pending_queue);
+	if (!skb) {
+		return 0;
+	}
+	lan91c111_send_hw(skb);
+
+	while (NULL != (skb = skb_queue_pop(&smc->pending_queue))) {
+		ret = lan91c111_tx_try(smc, skb);
+		if (ret) {
+			return ret;
+		}
+		skb_free(skb);
+	}
+
+	return 0;
+}
+
+static int lan91c111_recv_hw(struct net_device *dev_id) {
+	int i, packet;
 	struct sk_buff *skb;
 	uint32_t buf;
 	uint16_t len;
 	uint8_t *skb_data;
-	uint16_t int_status;
-	int i, packet;
-	ipl_t ipl;
 
-	ipl = ipl_save();
-	{
-		lan91c111_set_bank(2);
+	packet = (REG16_LOAD(BANK_FIFO_PORTS) >> 8) & 0xFF;
+	if (packet == FIFO_EMPTY) {
+		return 0;
+	}
 
-		int_status = REG16_LOAD(BANK_INTERRUPT_STATUS);
+	REG16_STORE(BANK_POINTER,
+			RX_FIFO_PACKET | AUTO_INCR | LAN91C111_STATUS_BYTES);
+	len = (REG16_LOAD(BANK_DATA) & PLEN_MASK) - 10;
+	/* In original structure, byte count includes headers, so
+	 * we shrink it to data size */
 
-		if (int_status & TX_INT) {
-			/* Clear TX interrupt */
-			int_status &= ~TX_INT;
-			REG8_STORE(BANK_INTERRUPT_STATUS, int_status);
-		}
+	skb = skb_alloc(len + LAN91C111_STATUS_BYTES);
+	assert(skb);
+	skb->len = len;
+	skb->dev = dev_id;
 
-		if (!(int_status & RX_INT)) {
-			goto out;
-		}
+	skb_data = skb_data_cast_in(skb->data);
+	assert(skb_data);
 
-		packet = (REG16_LOAD(BANK_FIFO_PORTS) >> 8) & 0xFF;
-		if (packet == FIFO_EMPTY) {
-			goto out;
-		}
+	for (i = 0; i < len >> 2; i++) {
+		buf = REG32_LOAD(BANK_DATA);
+		*((uint32_t *)(skb_data + i * 4)) = buf;
+	}
 
-		REG16_STORE(BANK_POINTER,
-				RX_FIFO_PACKET | AUTO_INCR | LAN91C111_STATUS_BYTES);
-		len = (REG16_LOAD(BANK_DATA) & PLEN_MASK) - 10;
-		/* In original structure, byte count includes headers, so
-		 * we shrink it to data size */
+	if (len % 4 == 2) {
+		buf = REG16_LOAD(BANK_DATA);
+		*((uint16_t *)(skb_data + i * 4)) = buf & 0xFFFF;
+	}
 
-		skb = skb_alloc(len + LAN91C111_STATUS_BYTES);
-		assert(skb);
-		skb->len = len;
-		skb->dev = dev_id;
+	/* Skip 4 bytes CRC */
+	buf = REG32_LOAD(BANK_DATA);
+	buf = REG16_LOAD(BANK_DATA);
 
-		skb_data = skb_data_cast_in(skb->data);
-		assert(skb_data);
+	lan91c111_set_cmd(CMD_RX_POP_AND_RELEASE);
 
-		for (i = 0; i < len >> 2; i++) {
-			buf = REG32_LOAD(BANK_DATA);
-			*((uint32_t *)(skb_data + i * 4)) = buf;
-		}
+	if (buf & (ODD_CONTROL << 8)) {
+		skb->len++;
```

Comment:
Are you sure it's neccessary? This also may cause buffer overflow.

Reply:
this code just move from old file line 315 to own routine lan91c111_recv_hw